llama-rs/src/convert.rs

```diff
@@ -128,13 +129,20 @@ fn write_header(fout: &mut File, hparams: &Hyperparameters) -> Result<(), String
 fn write_tokens(file: &mut File, vocab: &Vocabulary) -> Result<(), String> {
     let mut values: Vec<u8> = vec![];
     for (i, token) in vocab.id_to_token.iter().enumerate() {
-        let text = match token {
-            _ if token.contains("<unk>") => " \u{2047} ".as_bytes().to_vec(),
-            _ if token.contains("s>") => vec![],
-            _ if token.len() == 6 && token.contains("<0x") => {
-                vec![u8::from_str_radix(&token[3..5], 16).unwrap()]
+        // TODO: Not sure what the behaviour should be if the token is not valid UTF-8.
+        //
+        // Switching to the HF tokenizer should fix this.
```

Comment:
There shouldn't be any non-utf8 tokens here. Any non-utf8 is already represented through the byte fallback (i.e. the `<0x...>`) tokens, which we handle here.

I've seen two kinds of behaviors from llama.cpp here, one is to write the unicode replacement character �. The other is simply to dump invalid UTF-8 into the file. That said, this was in relation to byte fallback tokens and has since been fixed, so I'm not aware of any other situation where invalid UTF-8 would be stored in the vocabulary other tha a corrupt / malformed model file.

I think a warning here would be justified :thinking: